src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs

```diff
@@ -383,49 +377,68 @@ protected override void Dispose(bool disposing)
+        private sealed class RequestRegistration : TaskCompletionSource<RateLimitLease>
         {
-            public RequestRegistration(int requestedCount, TaskCompletionSource<RateLimitLease> tcs,
-                CancellationTokenRegistration cancellationTokenRegistration)
-            {
-                Count = requestedCount;
-                // Perf: Use AsyncOperation<TResult> instead
-                Tcs = tcs;
-                CancellationTokenRegistration = cancellationTokenRegistration;
-            }
+            private readonly CancellationToken _cancellationToken;
+            private CancellationTokenRegistration _cancellationTokenRegistration;
 
-            public int Count { get; }
+            // this field is used only by the disposal mechanics and never shared between threads
+            private RequestRegistration? _next;
 
-            public TaskCompletionSource<RateLimitLease> Tcs { get; }
+            public RequestRegistration(int permitCount, ConcurrencyLimiter limiter, CancellationToken cancellationToken)
+                : base(limiter, TaskCreationOptions.RunContinuationsAsynchronously)
+            {
+                Count = permitCount;
+                _cancellationToken = cancellationToken;
 
-            public CancellationTokenRegistration CancellationTokenRegistration { get; }
-        }
+                // RequestRegistration objects are created while the limiter lock is held
+                // if cancellationToken fires before or while the lock is held, UnsafeRegister
+                // is going to invoke the callback synchronously, but this does not create
+                // a deadlock because lock are reentrant
+                if (cancellationToken.CanBeCanceled)
+#if NETCOREAPP || NETSTANDARD2_1
+                    _cancellationTokenRegistration = cancellationToken.UnsafeRegister(Cancel, this) ;
+#else
+                    _cancellationTokenRegistration = cancellationToken.Register(Cancel, this) ;
+#endif
+            }
 
-        private sealed class CancelQueueState : TaskCompletionSource<RateLimitLease>
-        {
-            private readonly int _permitCount;
-            private readonly ConcurrencyLimiter _limiter;
-            private readonly CancellationToken _cancellationToken;
+            public int Count { get; }
 
-            public CancelQueueState(int permitCount, ConcurrencyLimiter limiter, CancellationToken cancellationToken)
-                : base(TaskCreationOptions.RunContinuationsAsynchronously)
+            private static void Cancel(object? state)
             {
-                _permitCount = permitCount;
-                _limiter = limiter;
-                _cancellationToken = cancellationToken;
+                if (state is RequestRegistration registration && registration.TrySetCanceled(registration._cancellationToken))
+                {
+                    var limiter = (ConcurrencyLimiter)registration.Task.AsyncState!;
```

Comment:
Never seen this before, kind of interesting since you can "avoid" the extra field. Not really avoiding really since it's there whether you use it or not 😄

Reply:
Yup, might as well use it and save a pointer.